addons/binding/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/UPBBaseHandler.java

```diff
@@ -0,0 +1,149 @@
+
+    private StateConverter stateConverter = new StateConverter();
+    private UPBMessageSender messageSender;
+    private byte id;
+
+    /**
+     * Instantiates a new {@link UPBBaseHandler}.
+     *
+     * @param thing the thing to be handled
+     */
+    public UPBBaseHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        this.id = ((Number) getConfig().get(UPBBindingConstants.DEVICE_ID)).byteValue();
+
+        super.initialize();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        byte[] commandBytes = getCommandBytes(channelUID, command);
+
+        if (messageSender == null) {
+            logger.warn("messageSender is null, command ignored");
+            return;
+        }
+
+        if (commandBytes == null) {
+            logger.warn("Failed to interpret command [{}]", command);
+            return;
+        }
+
+        MessageBuilder message = MessageBuilder.create().destination(getId()).link(isLink()).command(commandBytes);
+
+        if (command == RefreshType.REFRESH) {
+            message.priority(UPBMessage.Priority.LOW);
+        }
+
+        messageSender.sendMessage(message);
+    }
+
+    @Override
+    public void messageReceived(UPBMessage message) {
+        if (message.getDestination() != getId() && message.getSource() != getId()) {
+            return;
+        }
+
+        if (message.getControlWord().isLink() != isLink()) {
+            return;
+        }
+
+        updateStatus(ThingStatus.ONLINE);
+
+        // parse status
+        for (Channel c : getThing().getChannels()) {
+            State newState = stateConverter.convert(message, c.getUID().getId());
+
+            if (newState != null) {
+                updateState(c.getUID(), newState);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Converts a {@link Command} into an array of bytes to be passed to the UPB modem.
+     * 
+     * @param channelUID the channel to get the command for
+     * @param command the command to convert to byte form
+     * @return an array of bytes to send to the UPB modem, or null if the command could not be converted
+     */
+    protected byte[] getCommandBytes(ChannelUID channelUID, Command command) {
+        byte[] commandByte = null;
+
+        if (command == OnOffType.ON) {
+            commandByte = new byte[] { UPBMessage.Command.ACTIVATE.toByte() };
+        } else if (command == OnOffType.OFF) {
+            commandByte = new byte[] { UPBMessage.Command.DEACTIVATE.toByte() };
+        } else if (command instanceof PercentType) {
+            commandByte = new byte[] { UPBMessage.Command.GOTO.toByte(), ((PercentType) command).byteValue() };
+        } else if (command == RefreshType.REFRESH && !isLink()) {
+            commandByte = new byte[] { UPBMessage.Command.REPORT_STATE.toByte() };
+        }
```

Comment:
Maybe then add at least a debug log, that we received and 'ignored' a message. It could come in handy when people miss messages and try to find out what is going on.